### ISSUE TITLE
not consistenly getting "User Account Request"s

### ISSUE THREAD

**Issue by JChipault** [USGS-WiM/whispersservices_django#371]
We had a user (jchung) register as a Natural Resource Management Professional on the test site yesterday and whispers@usgs.gov never received a "User Account Request" email. 

In recent times, we've received notifications about account requests from DottieJohnson4 and apatel, but none for jchung, jctest, or dtunseth. It's spotty for some reason

**Comment by aaronstephenson** [comment]
I just tested this via the API on my laptop, on our dev server, and on the test sever, and it works as expected in each place. I see the new user change request record in that table every time. This suggests to me that the back end code is fine. @BlakeDraper could there be something happening on the front end, where the new user's change request gets dropped before POSTing to the API?

**Comment by BlakeDraper** [comment]
I'll take a look.

**Comment by BlakeDraper** [comment]
@aaronstephenson this is the format being sent. is this correct?

```
{
  "username": "371test2",
  "first_name": "371",
  "last_name": "test",
  "email": "371+2@me.com",
  "password": "56gh01@yasyas",
  "new_user_change_request": {
    "role_requested": 5,
    "organization_requested": 1737,
    "comment": "test test test"
  }
}
```

**Comment by aaronstephenson** [comment]
Yes. Here's what I've been using, they look identical in form:
```
{
    "username": "someguy24",
    "password": "abcdefg1234567!",
    "first_name": "Some",
    "last_name": "Guy",
    "email": "someguy24@example.biz",
    "new_user_change_request": {
	"role_requested": 5,
	"organization_requested": 1066,
	"comment": "Hello yes please."
    }
}
```

**Comment by BlakeDraper** [comment]
@JChipault sorry to have to ask this, but do you have any more details of these requests that have not behaved properly? namely: the org and the role requested? I can't believe that would have anything to do with it, but trying to explore all avenues because my testing is showing the requests go through.

**Comment by nbaertlein** [comment]
The user 'jchung' requested the 'Partner' role (id = 5)  for the DHS organization (ID = 1756).  This request was logged into the table whispers_userchangerequest (see SQL below).  When a user requests to be partner we (the org and admins) also typically get a notification for a 'User Account Change Request', but in this case we did not (see SQL below).  We saw similar behavior for users jctest and dtunseth earlier in the month, but these did not log into whispers_userchangerequest either.

I'm assuming both these actions (request log and notification) are triggered from the same action of requesting to be a partner (cascading or simultaneously), but please correct me if I'm wrong.  As you indicated, it appears to be working most of the time, so maybe we wait and see given our time constraints.  I just don't want to miss a request if possible.

**whispers_userchangerequest for a user:**
select a.id as user_id, a.username, a.date_joined,a.role_id, b.*
from public.whispers_user a 
	left join public.whispersservices_historicaluserchangerequest b on a.id = b.created_by_id
where a.username = 'jchung'
order by a.date_joined desc, b.id desc, b.history_id desc;

**whispers_notification for a account change request:**
select a.id as user_id, a.username, a.date_joined, a.role_id, b.*
from public.whispers_user a left join
	(select * from public.whispers_notification where subject like '%User Account Change Request%') b on a.username = b.source
where a.username = 'jchung'
order by a.date_joined desc, b.id desc;

**Comment by nbaertlein** [comment]
...also regarding 'We saw similar behavior for users jctest and dtunseth earlier in the month, but these did not log into whispers_userchangerequest either.'  I think logging in whispers_userchangerequest was resolved around 4/7.

**Comment by BlakeDraper** [comment]
@nbaertlein there were updates to this part of the code recently, so as you say, most of that previous behavior is no longer relevant. Would you mind just creating a few dummy accounts through the normal frontend process and see if the behavior is as expected?  you could then just delete those accounts after creating them with your data admin privileges.

**Comment by nbaertlein** [comment]
I just created 5 dummy accounts with different combinations of partner and organization, including leaving blank.  I found no error.  I think we'll have to let it ride and just keep an eye on it.

**Comment by nbaertlein** [comment]
Circling back around on this one.  4 accounts were created yesterday, but we received the change request notification on only 2.  Attached is a spreadsheet of user accounts created since 4/16, with sql code, that links 3 tables: user, userchangerequest, and notification ("User Account Change Request" only).  Most are notifying correctly, 3 have missed and are highlighted in orange.
[Partner_requests_may2_20200512.xlsx](https://github.com/USGS-WiM/whispersservices_django/files/4616755/Partner_requests_may2_20200512.xlsx)

**Comment by aaronstephenson** [comment]
Thanks @nbaertlein. Do you happen to know the requested role IDs of the 3 that failed?

**Comment by nbaertlein** [comment]
From newest to oldest, the requested role IDs were 3, 4, and 5.

**Comment by aaronstephenson** [comment]
Thanks @nbaertlein, I found the source of the problem. The celery logs show an error during the notification creation task:
```django.db.utils.DataError: value too long for type character varying(128)```
which is in reference to the `subject` field, which has a max length of 128 characters.

The 3 failures you report came from 2 organizations which, combined with the boilerplate prefix for the subject of that notification ("User Account Change Request: "), results in a string that exceeds the limit for that field:
`User Account Change Request: Southeastern Cooperative Wildlife Disease Study - College of Veterinary Medicine - University of Georgia` (133 chars)
`User Account Change Request: Florida Fish and Wildlife Conservation Commission - Fish and Wildlife Research Institute - Gainesville` (131 chars)

The simplest solution would be to manually shorten the organization names in the Organization table, because we wouldn't need to make any changes to the code.

Otherwise our options would all involve code changes: either force trim the subject line to 128 chars (potentially resulting in confusing org names in the subject line), or increase the max length of the `subject` field to maybe 256 (would need to coordinate with Blake to ensure this doesn't break anything on the front end), or both.

Let me know how you would like to proceed.

**Comment by nbaertlein** [comment]
That makes sense.   We'll manage it from the data side so no code changes needed.  Are any subject templates referred to in code?  Wondering if we could also shorten "User Account Change Request" text in the template table to something like "Account request".

@JChipault We'll need to shorten a few of the org names.  At minimum, to avoid the issue above, we'll need to trim names longer than 100 characters- currently there are 22.

**Comment by aaronstephenson** [comment]
I went ahead and added a snippet of code in the notification generator to check for overlong subject lines, and email the NWHC admins with details if that happens, rather than silently fail when encountering that error.

Implemented in ae5d77b4
